File: src/java/prbm_stackqueue/Top.java
package prbm_stackqueue;

import java.util.Stack;

public class Top {

	public static void main(String[] args) {
		int[] heights = {3, 9, 9, 3, 5, 7, 2};
		for(int index : solution(heights)) {
			System.out.print(index+" ");
		}		
		
	}
    public static int[] solution(int[] heights) {
    	// 역으로 loop를 돌면되겠는데.. 이게 stack / queue인 이유는?
    	int[] answer = new int[heights.length];
    	for(int i=heights.length-1; i>0; i--) {
    		int receiver = 0;
    		for(int j=i-1; j>=0; j--) {
    			if(heights[j]>heights[i]) {
    				answer[i]=j+1;
    				break; // break안해줘서 처음에 틀림. j+1도 index값이 0이 아닌 1부터 시작해야 함.
    			}
    		}
    	}
        return answer;
    }
    
    // Stack을 사용한 풀이
    class Tower {
        int idx;
        int height;

        public Tower(int idx, int height) {
            this.idx = idx;
            this.height = height;
        }
    }

    public int[] solution2(int[] heights) {
        Stack<Tower> st = new Stack<>();
        int[] answer = new int[heights.length];

        // 여기서는 loop를 정방향으로 도는게 특징!
        for (int i = 0; i < heights.length; i++) {
            Tower tower = new Tower(i + 1, heights[i]);
            int receiveIdx = 0;

            while (!st.isEmpty()) {
                Tower top = st.peek();

                if (top.height > tower.height) {
                    receiveIdx = top.idx;
                    break;
                }

                st.pop();
            }

            st.push(tower);
            answer[i] = receiveIdx;
        }

        return answer;
    }
}

File: src/py/pg/prbm_stackqueue/Top.py
def solution(heights):
    stack = []
    answer = []
    for i, h in enumerate(heights) :
        print(stack)
        print(answer)

        if len(stack)==0 :
            stack.append(i)
            answer.append(0)
        elif heights[stack[len(stack)-1]]>h :
            answer.append(stack[len(stack)-1]+1)            
            stack.append(i)            
        elif heights[stack[len(stack)-1]]<=h :
            while len(stack)>0 and heights[stack[len(stack)-1]]<=h :
                stack.pop()
            if len(stack)==0 :
                answer.append(0)
            else :
                answer.append(stack[len(stack)-1]+1)            
            stack.append(i)    
    return answer

## 훨씬 깔끔한 풀이
def solution2(heights):
    stack = []
    answer = []

    for i, h in enumerate(heights) :
        receiver_idx = 0

        while len(stack)>0 :
            top_idx = stack[len(stack)-1]

            if heights[top_idx] > h :
                receiver_idx = top_idx+1
                break
            stack.pop()

        stack.append(i)
        answer.append(receiver_idx)
        
    return answer